### DISCUSSION TITLE
thread pools and memory usage

### DISCUSSION THREAD

**Discussion by sbfaulkner** [libvips/libvips#2552] [Q&A] [answered]
I've been digging into an issue related to libvips memory usage for an image-processing service my team works on

We noticed high memory usage for a frame-by-frame resizing of an animated gif

ie.
```
for each page {
  extract_area -> premultiply -> resize -> unpremultiply
}

join
```

I was surprised to observe that concurrency increased memory usage...
ie.
```
VIPS_LEAK=1 VIPS_CONCURRENCY=1 VIPS_PROGRESS=1 ./gifresize ../yellow.gif
gifresize temp-2928: 360 x 31486 pixels, 1 threads, 360 x 16 tiles, 128 lines in buffer
gifresize (null): 1350 x 118300 pixels, 1 threads, 1350 x 16 tiles, 128 lines in buffer
gifresize (null): done in 1.59s
gifresize temp-2928: done in 8.68s
memory: high-water mark 1.63 GB

VIPS_LEAK=1 VIPS_CONCURRENCY=2 VIPS_PROGRESS=1 ./gifresize ../yellow.gif
gifresize temp-2928: 360 x 31486 pixels, 2 threads, 360 x 16 tiles, 128 lines in buffer
gifresize (null): 1350 x 118300 pixels, 2 threads, 1350 x 16 tiles, 128 lines in buffer
gifresize (null): done in 1.68s
gifresize temp-2928: done in 6.15s
memory: high-water mark 3.25 GB

VIPS_LEAK=1 VIPS_CONCURRENCY=3 VIPS_PROGRESS=1 ./gifresize ../yellow.gif
gifresize temp-2928: 360 x 31486 pixels, 3 threads, 360 x 16 tiles, 128 lines in buffer
gifresize (null): 1350 x 118300 pixels, 3 threads, 1350 x 16 tiles, 128 lines in buffer
gifresize (null): done in 1.58s
gifresize temp-2928: done in 5.5s
memory: high-water mark 4.88 GB

VIPS_LEAK=1 VIPS_CONCURRENCY=4 VIPS_PROGRESS=1 ./gifresize ../yellow.gif
gifresize temp-2928: 360 x 31486 pixels, 4 threads, 360 x 16 tiles, 128 lines in buffer
gifresize (null): 1350 x 118300 pixels, 4 threads, 1350 x 16 tiles, 128 lines in buffer
gifresize (null): done in 1.54s
gifresize temp-2928: done in 5.39s
memory: high-water mark 6.50 GB

VIPS_LEAK=1 VIPS_CONCURRENCY=5 VIPS_PROGRESS=1 ./gifresize ../yellow.gif
gifresize temp-2928: 360 x 31486 pixels, 5 threads, 360 x 16 tiles, 128 lines in buffer
gifresize (null): 1350 x 118300 pixels, 5 threads, 1350 x 16 tiles, 128 lines in buffer
gifresize (null): done in 1.48s
gifresize temp-2928: done in 5.29s
memory: high-water mark 8.12 GB
```

### questions
1. is it expected that memory usage increase with concurrency?
2. followup question(s)... if we're invoking libvips from multiple threads in our main program, how does concurrency work? does each image get dedicated threads? or do multiple concurrent requests contend for a single shared thread pool?

**Comment by jcupitt** [comment, marked as answer]
Hi @sbfaulkner,

You don't need to resize frame by frame -- try eg.:

```
vipsthumbnail silver-gear-cogs-animation-5.gif[n=-1] --size 128x128 -o x.gif
```

It should be quite a bit quicker than working in frames. 

libvips 8.12 has a new GIF writer that ought to be maybe 2x faster and need up to 10x less memory. There are some benchmarks here:

https://github.com/libvips/libvips/pull/2445#issuecomment-932751529

"per frame" is 8.12 behaviour. You can see memuse is low and steady with any GIF file. The 8.11 saver was based on imagemagick and while quality was good, speed and memory use was not.

Memory use scales with image width and with concurrency. libvips pipelines are split into immutable and mutable parts, and each thread gets a copy of the mutable bits. It depends what's in the pipeline, but there's not always enough parallelism available for threads to be worthwhile. I'd benchmark and limit pipeline parallelism to what's beneficial, and no more.

There's one libvips worker pool per process, but it will expand as required. The concurrency setting sets the per-pipeline pool size.

**Comment by sbfaulkner** [comment]
hey @jcupitt 

yes, I we are changing the way the processing works to use thumbnail (or at least resize) and are looking forward to reaping the benefits of the libcgif support

thanks for the other details ... tl;dr -- yes the proportionate memory usage is expected, and the thread pooling is per-pipeline
